### PR TITLE
Fix #2063: introduce :defined-by->lint-as + always keep original in :defined-by

### LINKED ISSUE #2063: `:defined-by` in analysis output uses `:lint-as` rather than symbol itself
[ Leave this text in if you want to promote it. ]

To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

[ To keep development of this project going, consider sponsoring. If you are
already a sponsor, thank you! You can remove this message. ]

**Is your feature request related to a problem? Please describe.**

I'm trying to find uses of a particular var, when defined inside another custom type of var.  In this case, a `is` within `defspec` (which isn't valid, and I want to find out how rampant this problem is).

**Describe the solution you'd like**

`:defined-by` shouldn't use `nil` for `lint.as/def-catch-all`, nor should it copy the value from `:lint-as` to `:defined-by`.

That is, for:

```clojure
{:lint-as {ns/a clj-kondo.lint-as/def-catch-all ns/b clojure.core/def}}
```

`:defined-by` should be `ns/a` and `ns/b` respectively for any uses.  Rather than `nil` and `clojure.core/def` as it is now.

**Describe alternatives you've considered**

I'll probably take another stab at solving my problem with grasp and postwalk.

**Additional context**

https://clojurians.slack.com/archives/CHY97NXE2/p1683106125566899

### PR DESCRIPTION
**Pull request by borkdude**

### CODE REVIEW

**Comment by borkdude** [comment]
@ericdallo Can you review this and also test on clojure-lsp?

File: src/clj_kondo/impl/analysis.clj [hunk 1]
```diff
@@ -187,6 +189,7 @@
                 :impl-ns impl-ns
                 :filename filename
                 :defined-by defined-by
+                :defined-by->lint-as defined-by->lint-as
```
Comment (ericdallo) [inline src/clj_kondo/impl/analysis.clj:192]: This formatting looks odd, too much extra spaces?

Comment (ericdallo) [reply src/clj_kondo/impl/analysis.clj:192]: There is the same in other places, I suppose you used some find-replace?

**Review by ericdallo — APPROVED** [review]
Looks good!
I will only have time to test with LSP in the weekend unfortunately

File: src/clj_kondo/impl/analysis.clj (hunk 1 above)
Comment (borkdude) [reply src/clj_kondo/impl/analysis.clj:192]: I'm not sure what you mean

**Pull request merged**